risc0/zkp/core/src/sha.rs

```diff
@@ -12,101 +12,64 @@
+        Digest(data)
+    }
 
-fn set_word(buf: &mut [u8], idx: usize, word: u32) {
-    buf[(4 * idx)..(4 * idx + 4)].copy_from_slice(&word.to_be_bytes());
-}
+    pub fn try_from_slice(words: &[u32]) -> Result<Self> {
+        Ok(Digest(words.try_into()?))
+    }
 
-impl Digest {
-    pub fn from_u32s(words: &[u32]) -> Self {
-        Digest(words.try_into().expect("slice with incorrect length"))
+    pub fn from_slice(words: &[u32]) -> Self {
+        Self::try_from_slice(words).unwrap()
     }
 
-    // Digests a byte array in the bog-standard SHA-256 fashion, but returns
-    // the resulting value as our digest type (8 words) rather than 32 bytes.
-    pub fn hash_bytes(bytes: &[u8]) -> Self {
-        let mut hasher = Sha256::new();
-        hasher.update(bytes);
-        Digest(
-            hasher
-                .finalize()
-                .as_slice()
-                .chunks(4)
-                .map(|chunk| u32::from_be_bytes(chunk.try_into().unwrap()))
-                .collect::<Vec<u32>>()
-                .try_into()
-                .unwrap(),
-        )
+    pub fn as_slice(&self) -> &[u32] {
+        &self.0
     }
 
-    // Digest FPs a word at a time, without padding (and using le encoding)
-    pub fn hash_fps(fps: &[Fp]) -> Self {
-        let mut state = INIT_256;
-        let mut block: GenericArray<u8, U64> = GenericArray::default();
-        let mut off = 0;
-        for i in 0..fps.len() {
-            set_word(block.as_mut_slice(), off, u32::from(fps[i]));
-            off += 1;
-            if off == 16 {
-                compress256(&mut state, slice::from_ref(&block));
-                off = 0;
-            }
-        }
-        if off != 0 {
-            block[off * 4..].fill(0);
-            compress256(&mut state, slice::from_ref(&block));
-        }
-        Digest(state)
+    pub fn as_mut_slice(&mut self) -> &mut [u32] {
+        &mut self.0
     }
 
-    pub fn hash_fp4s(fp4s: &[Fp4]) -> Self {
-        let mut flat: Vec<Fp> = vec![];
-        for i in 0..fp4s.len() {
-            for j in 0..4 {
-                flat.push(fp4s[i].elems()[j]);
-            }
-        }
-        return Self::hash_fps(&flat);
+    pub fn get(&self) -> &[u32; DIGEST_WORDS] {
+        &self.0
     }
 
-    // Digest two digest into one
-    pub fn hash_pair(a: &Digest, b: &Digest) -> Self {
-        let mut state = INIT_256;
-        let mut block: GenericArray<u8, U64> = GenericArray::default();
-        for i in 0..8 {
-            set_word(block.as_mut_slice(), i, a.0[i]);
-            set_word(block.as_mut_slice(), 8 + i, b.0[i]);
-        }
-        compress256(&mut state, slice::from_ref(&block));
-        Digest(state)
+    pub fn get_mut(&mut self) -> &mut [u32; DIGEST_WORDS] {
+        &mut self.0
     }
-}
 
-impl Digest {
-    /// Constructs a new `Digest` from a byte array.
-    pub fn new(data: [u32; DIGEST_WORDS]) -> Self {
-        Self(data)
+    pub fn to_hex(&self) -> String {
+        format!("{}", self)
```

Comment:
`format!` is actually pretty expensive in the guest, it causes a large amount of bloat. I'm OK with this for now, but perhaps a TODO would be good.

Reply:
Done